File: Solutions/131-Palindrome Partitioning/131-Palindrome Partitioning.js
//Third fastest
const isPalindrome = (s, left, right) => {
    while (left < right) {
        if (s[left] !== s[right]) return false
        left++
        right--
    }
    return true
}

const recurse = (str, ptr, progress, valid) => {
    if (ptr === str.length) {
        valid.push([...progress])
    } else {
        for (let i=ptr; i<str.length; i++) {
            if (isPalindrome(str, ptr, i)) {
                let snippet = str.substring(ptr, i+1)
                progress.push(snippet)
                recurse(str, i+1, progress, valid)
                progress.pop()
            }
        }
    }
}

const partition = str => {
    let validDecomp = []
    let decompInProgress = []
    recurse(str, 0, decompInProgress, validDecomp)
    return validDecomp
}

//Second fastest
var partition = function(s) {
    const result = [];
    recurse(0, s, [], result);
    return result;
};

function recurse(start, s, currList, result) {
    if(start >= s.length) { 
        result.push(Array.from(currList));
    }
    for(let end = start; end < s.length; end++) {
        if(isPalin(s, start, end)) {
            currList.push(s.substring(start, end+1));
            recurse(end+1, s, currList, result);
            currList.pop();
        }
    }
}

function isPalin(s, low, hi) {
    while(low < hi) {
        if(s.charAt(low) != s.charAt(hi)) {
            return false;
        }
        low++;
        hi--;
    }
    return true;
}

//Fastest
var partition = function(s) { const res = new Set();
    if (s.length === 0) return [];
    
    const isPal = (start, end) => {
        while (start<=end) {
            if(s[start] !== s[end]) return false;
            
            start++;
            end--;
        }
        
        return true;
    }
    
    const helper = (position, cur) => {
        if(position === s.length) {
            res.add([...cur]);
            return;
        }
        
        for(let i=position; i<s.length;i++) {
            if(isPal(position, i)) {
                cur.push(s.substring(position, i+1));
                helper(i+1, cur);
                cur.pop();
            }
        }
    }
    
    helper(0, []);
    
    return [...res];
};

File: Solutions/131-Palindrome Partitioning/131-Palindrome Partitioning.go
//Third fastest
func isPalindrome(s string) bool {
    for i, j := 0, len(s)-1; i < j; i,j = i+1,j-1 {
        if s[i] != s[j] {
            return false
        }
    }
    
    return true
}

func partition(s string) [][]string {
    if len(s) == 0 {
        return [][]string{{}}
    }
    
    res := [][]string{}
    for i := 0; i < len(s); i += 1 {
        cur := s[:i+1]
        if !isPalindrome(cur) {
            continue
        }
        rest := partition(s[i+1:])
        
        for _, x := range(rest) {
            t := append([]string{cur}, x...)
            res = append(res, t)
        }
    }
    
    return res
}

//Second fastest
func partition(s string) [][]string {
    
    results := make([][]string, 0)
    collect(s, []string{}, &results)
    return results
    
    
}

func collect(s string, current []string, results *[][]string) {
    if len(s) == 0 {
        *results = append(*results, current)
        return
    }
    
    for i := 1; i <= len(s); i++ {
        if s[0] == s[i-1] && isPalin(s[:i]) {
            collect(s[i:], append(append([]string{}, current...), s[:i]), results)
        }
    }
    
}

func isPalin(s string) bool {
    left, right := 0, len(s) - 1
    for left < right {
        if s[left] != s[right] {
            return false
        }
        left++
        right--
    }
    
    return true
}

//Fastest
func partition(s string) [][]string {
    
	var ret [][]string
	var cur []string

	helper(cur, s, &ret)

	return ret
}

func helper(cur []string, s string, ret *[][]string) {
	p := len(strings.Join(cur, ""))
	if p == len(s) {
		tmp := make([]string,len(cur))
		copy(tmp,cur)
		*ret = append(*ret, tmp)
		return
	}
	for i := p; i < len(s); i++ {
		l := p
		r := i
		if isPalindrome(s,l,r) {
			helper(append(cur, s[l:r+1]), s, ret)
		}

	}
}

func isPalindrome(s string, l int, r int) bool{
	for l < r {
		if s[l] != s[r] {
			return false
		}
		l++
		r--
	}

	return true
}

//